core/src/main/java/org/ehcache/events/CacheEventNotificationServiceImpl.java

```diff
@@ -174,60 +243,89 @@ public boolean hasListeners() {
-          return CacheEventListenerProvider.class;
-        }
-        
-        @Override
-        public EventOrdering orderingMode() {
-          return ordering;
-        }
-        
-        @Override
-        public EventFiring firingMode() {
-          return firing;
-        }
-        
-        @Override
-        public EnumSet<EventType> fireOn() {
-          return forEvents;
-        }
-      };
+  private final Set<CacheEventListener> orderedListenersSet = new CopyOnWriteArraySet<CacheEventListener>();
+  private final Set<EventListenerWrapper> syncListenersSet = new CopyOnWriteArraySet<EventListenerWrapper>();
+  private final Set<EventListenerWrapper> aSyncListenersSet = new CopyOnWriteArraySet<EventListenerWrapper>();
+
+  private boolean addOrderedListenerToSet(CacheEventListener<? super K, ? super V> listener, EventOrdering ordering) {
+    eventDispatcher = eventDispatchers.get(1);
+    return ordering != EventOrdering.ORDERED || orderedListenersSet.add(listener);
+  }
+  
+  private boolean removeOrderedListenerFromSet(CacheEventListener<? super K, ? super V> listener) {
+    orderedListenersSet.remove(listener);
+    if (orderedListenersSet.isEmpty()) {
+      eventDispatcher = eventDispatchers.get(0);
     }
-    
-    @SuppressWarnings("unchecked")
-    <K, V> CacheEventListener<K, V> getListener() {
-      return (CacheEventListener<K, V>) listener;
+    return true;
+  }
+
+  @Override
+  public List<CacheConfigurationChangeListener> getConfigurationChangeListeners() {
+    List<CacheConfigurationChangeListener> configurationChangeListenerList = new ArrayList<CacheConfigurationChangeListener>();
+    configurationChangeListenerList.add(new CacheConfigurationChangeListener() {
+      @Override
+      public void cacheConfigurationChange(final CacheConfigurationChangeEvent event) {
+        if (event.getProperty().equals(CacheConfigurationProperty.ADDLISTENER)) {
+          EventListenerWrapper newListener = (EventListenerWrapper)event.getNewValue();
+          CacheEventListenerConfiguration listenerConfiguration = newListener.config;
+          registerCacheEventListener(newListener.getListener(), listenerConfiguration.orderingMode(),
+              listenerConfiguration.firingMode(), listenerConfiguration.fireOn());
+        } else if (event.getProperty().equals(CacheConfigurationProperty.REMOVELISTENER)) {
+          CacheEventListener<K, V> oldListener = (CacheEventListener)event.getOldValue();
+          deregisterCacheEventListener(oldListener);
+        }
+      }
+    });
+    return configurationChangeListenerList;
+  }
+
+  @Override
+  public void fireEvents() {
+    while (!eventThreadLocal.get().isEmpty()) {
+      CacheEventWrapper cacheEventWrapper = eventThreadLocal.get().get(0); 
+      if (!cacheEventWrapper.hasFailed()) {
+        dispatchEvent(cacheEventWrapper);
+      }
+      eventThreadLocal.get().remove(0);
+      removeMapEntry((K)cacheEventWrapper.cacheEvent.getKey());
```

Comment:
I do not understand why you iterate over the whole event queue each time you process a single element.
According to the logic, you only need to remove events that are marked as failed. So you should add a single removal to an `else` on the `if` above.